src/core/ddsc/tests/psmxif.c

```diff
@@ -0,0 +1,441 @@
+
+#include "dds/dds.h"
+#include "dds/ddsi/ddsi_entity_index.h"
+#include "ddsi__addrset.h"
+#include "ddsi__entity.h"
+#include "ddsi__xevent.h"
+#include "dds__entity.h"
+#include "dds__serdata_default.h"
+
+#include "config_env.h"
+#include "test_common.h"
+#include "psmx_dummy_public.h"
+#include "Array100.h"
+#include "DynamicData.h"
+#include "PsmxDataModels.h"
+
+static void free_strings(uint32_t len, char** strings)
+{
+    if (len != 0 && strings != NULL) {
+        for (uint32_t i = 0; i < len; i++) {
+            dds_free(strings[i]);
+        }
+    }
+    dds_free(strings);
+}
+
+/**
+ * @brief Create a participant object, providing a configuration to choose the psmx interface.
+ * 
+ * @param[in] int_dom the domain id
+ * @param[in] cdds_psmx_name the name of the psmx interface to use
+ * @param[in] locator an array of 16 bytes, NULL if not used
+ * @return dds_entity_t the participant
+ */
+static char* create_config(dds_domainid_t int_dom, const char* cdds_psmx_name, const uint8_t* locator)
+{
+  char *configstr;
+  char locator_str[74];
+  if ( locator == NULL ) {
+    locator_str[0] = 0;
+  } else {
+    const uint8_t *l = locator;
+    snprintf(
+      locator_str,
+      sizeof(locator_str),
+      "LOCATOR=%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x;",
+      l[0], l[1], l[2], l[3], l[4], l[5], l[6], l[7], l[8], l[9], l[10], l[11], l[12], l[13], l[14], l[15]
+    );
+  }
+  ddsrt_asprintf (&configstr, "\
+${CYCLONEDDS_URI}${CYCLONEDDS_URI:+,}\
+<General>\
+  <AllowMulticast>spdp</AllowMulticast>\
+  <Interfaces>\
+    <PubSubMessageExchange name=\"%s\" library=\"psmx_%s\" priority=\"1000000\" config=\"%sSERVICE_NAME=psmx%d;KEYED_TOPICS=true;\" />\
+  </Interfaces>\
+</General>\
+<Discovery>\
+  <Tag>${CYCLONEDDS_PID}</Tag>\
+  <ExternalDomainId>0</ExternalDomainId>\
+</Discovery>\
+<Tracing>\
+  <OutputFile>cdds.log.%d</OutputFile>\
+</Tracing>\
+",
+    cdds_psmx_name,
+    cdds_psmx_name,
+    locator_str,
+    locator ? (int) locator[0] : -1, // This prevents plugins from forwarding across the "network".
+    (int) int_dom // log file name
+  );
+  char *xconfigstr = ddsrt_expand_envvars (configstr, int_dom);
+  ddsrt_free (configstr);
+  return xconfigstr;
+}
+
+static bool endpoint_has_psmx_enabled (dds_entity_t rd_or_wr)
+{
+  dds_return_t rc;
+  struct dds_entity *x;
+  bool psmx_enabled = false;
+  rc = dds_entity_pin (rd_or_wr, &x);
+  CU_ASSERT_FATAL (rc == DDS_RETCODE_OK);
+  switch (dds_entity_kind (x))
+  {
+    case DDS_KIND_READER: {
+      struct dds_reader const * const rd = (struct dds_reader *) x;
+      psmx_enabled = (rd->m_endpoint.psmx_endpoints.length > 0);
+      break;
+    }
+    case DDS_KIND_WRITER: {
+      struct dds_writer const * const wr = (struct dds_writer *) x;
+      psmx_enabled = (wr->m_endpoint.psmx_endpoints.length > 0);
+      break;
+    }
+    default: {
+      CU_ASSERT_FATAL (dds_entity_kind (x) == DDS_KIND_READER || dds_entity_kind (x) == DDS_KIND_WRITER);
```

Comment:
Writing this assert kinda defeats the purpose of using a `switch` to begin with. Then it makes more sense to do `if{READER}else{CU_ASSERT_FATAL(WRITER)`

Reply:
This function was copied from `psmx.c`. I suppose then it should be modified in both places.